File: implementation/swift/GameEventScriptTool/Sources/GameEventScriptTool/RunConsole.swift
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

import Foundation
import GameEventScriptCompiler
import GameEventScriptRuntime

extension Tool {
    func runConsole(_ session: RunSession, options: RunOptions) throws -> Bool {
        if io.inputTerminal {
            io.line(
                "GES event console. Type :help for commands and examples, :load <file> to add a program, or :quit to exit.",
                toError: true)
        }
        let prompt =
            options.color && io.inputTerminal && io.outputTerminal && io.errorTerminal ? TerminalPrompt(io: io) : nil
        var success = true
        var lineNumber = 0
        while io.outputError == nil {
            if io.inputTerminal && prompt == nil { io.write("ges> ", toError: true) }
            guard var line = try prompt?.readLine() ?? (prompt == nil ? io.input() : nil) else { return success }
            lineNumber += 1
            if lineNumber == 1 && line.hasPrefix("\u{feff}") { line.removeFirst() }
            let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.isEmpty { continue }
            let end = trimmed.firstIndex(where: \.isWhitespace) ?? trimmed.endIndex
            let command = String(trimmed[..<end])
            let argument = String(trimmed[end...]).trimmingCharacters(in: .whitespacesAndNewlines)
            if command == ":quit" && argument.isEmpty { return success }
            if command == ":help" {
                if !consoleHelp(argument) { success = false }
                continue
            }
            if command == ":list" && argument.isEmpty {
                session.inventory.programs()
                continue
            }
            if command == ":handler" && argument.isEmpty {
                session.inventory.registeredHandlers()
                continue
            }
            if ![":load", ":dump", ":source", ":unload", ":unloadAll", ":reload"].contains(command),
                command.hasPrefix(":"),
                command.dropFirst().allSatisfy({ $0 >= "a" && $0 <= "z" })
            {
                io.line(
                    "error cli.consoleCommand: Unknown command or arguments '\(trimmed)'. Use :help.", toError: true)
                success = false
                continue
            }
            var instance: GameEventScriptInstance?
            var path = "<interactive:\(lineNumber)>"
            do {
                defer { instance?.detach() }
                if command == ":unload" {
                    try session.inventory.unload(argument)
                    if !options.quiet { io.line("Unloaded \(argument).", toError: true) }
                    continue
                }
                if command == ":unloadAll" || command == ":reload" {
                    guard argument.isEmpty else { throw ToolError.usage("\(command) accepts no arguments.") }
                    if command == ":unloadAll" {
                        let count = session.inventory.unloadAll()
                        if !options.quiet {
                            io.line("Unloaded \(count) programs. Native console handlers remain active.", toError: true)
                        }
                    } else {
                        guard try session.reload(activePath: &path) else { return false }
                        if !options.quiet {
                            io.line(
                                "Reloaded all active programs on a fresh host; initialization completed.", toError: true
                            )
                        }
                    }
                    continue
                }
                if command == ":dump" {
                    try session.inventory.dump(argument, color: options.color)
                    continue
                }
                if command == ":source" {
                    try session.inventory.source(argument, color: options.color)
                    continue
                }
                if command == ":load" {
                    path = try Self.loadPath(argument)
                    try session.inventory.load(ToolFiles.read(path), paths: [path])
                    guard try session.pump() else { return false }
                    if !options.quiet {
                        io.line("Loaded \(path); initialization completed. Handlers remain active.", toError: true)
                    }
                    continue
                }
                let program = try GameEventScriptBuilder().addScript(
                    "on initialization {\n" + line + "\n}\n", sourceName: path
                ).compile()
                let handlers = program.bindings.filter(isHandler)
                guard handlers.count == 1, program.stringConstants[Int(handlers[0].name)] == "initialization" else {
                    io.line(
                        "error cli.interactiveInput: Enter handler-body statements, not additional handler declarations.",
                        toError: true)
                    success = false
                    continue
                }
                instance = try session.host.load(program)
                guard try session.pump() else { return false }
            } catch {
                io.report(error, fallback: path, argumentCode: "cli.consoleCommand")
                success = false
            }
        }
        return false
    }
    func consoleHelp(_ topic: String) -> Bool {
        let text: String
        switch topic {
        case "": text = ToolHelp.console
        case "load", ":load": text = ToolHelp.load
        case "unload", ":unload", "unloadAll", ":unloadAll", "reload", ":reload": text = ToolHelp.lifecycle
        case "list", ":list", "handler", ":handler", "dump", ":dump", "source", ":source": text = ToolHelp.inspection
        default:
            io.line(
                "error cli.consoleCommand: Unknown help topic '\(topic)'. Use :help, :help load, or :help dump.",
                toError: true)
            return false
        }
        io.line(toError: true)
        io.line(text, toError: true)
        io.line(toError: true)
        return true
    }
    static func loadPath(_ text: String) throws -> String {
        var path = text
        if let quote = path.first, quote == "'" || quote == "\"" {
            guard path.count >= 2, path.last == quote else {
                throw ToolError.usage("The quoted :load path must end with a matching quote.")
            }
            path = String(path.dropFirst().dropLast()).replacingOccurrences(
                of: String(repeating: quote, count: 2), with: String(quote))
        }
        guard !path.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw ToolError.usage("Specify one file after :load. Use :help load for examples.")
        }
        _ = try ToolFiles.fullPath(path)
        return path
    }
}

File: implementation/csharp/GameEventScript.Tool/src/RunConsole.cs
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

using System.Text;
using GameEventScript.Api;

namespace GameEventScript.Tool;

internal static class RunConsole
{
    internal static bool Run(RunSession session, bool color, bool quiet)
    {
        var terminal = !Console.IsInputRedirected;
        if (terminal) Console.Error.WriteLine("GES event console. Type :help for commands and examples, :load <file> to add a program, or :quit to exit.");
        using var input = new StreamReader(Console.OpenStandardInput(), new UTF8Encoding(false, true), detectEncodingFromByteOrderMarks: false);
        var history = new RunHistory();
        using var prompt = color && terminal && !Console.IsOutputRedirected && !Console.IsErrorRedirected ? new RunPrompt(history) : null;
        var success = true;
        var lineNumber = 0;
        while (true)
        {
            var waiting = session.Host.NextMessageDelay is not null;
            if (terminal && prompt is null && !waiting) Console.Error.Write("ges> ");
            var line = waiting ? RunWaitingInput.Read(session, input, terminal && !Console.IsErrorRedirected, color, history) : prompt is null ? input.ReadLine() : prompt.ReadLine();
            if (line is null) return success;
            history.Add(line);
            lineNumber++;
            if (lineNumber == 1 && line.StartsWith('\uFEFF')) line = line[1..];
            if (string.IsNullOrWhiteSpace(line)) continue;
            var trimmed = line.Trim();
            var separator = 0;
            while (separator < trimmed.Length && !char.IsWhiteSpace(trimmed[separator])) separator++;
            var command = trimmed[..separator];
            var commandArgument = trimmed[separator..].Trim();
            if (command == ":quit" && commandArgument.Length == 0) return success;
            if (command == ":help")
            {
                if (!RunConsoleHelp.Write(commandArgument)) success = false;
                continue;
            }
            if (command == ":list" && commandArgument.Length == 0)
            {
                session.Inventory.WritePrograms();
                continue;
            }
            if (command == ":handler" && commandArgument.Length == 0)
            {
                session.Inventory.WriteHandlers();
                continue;
            }
            if (command is not (":load" or ":dump" or ":source" or ":unload" or ":unloadAll" or ":reload") && command.StartsWith(':') && command.AsSpan(1).IndexOfAnyExceptInRange('a', 'z') < 0)
            {
                Console.Error.WriteLine($"error cli.consoleCommand: Unknown command or arguments '{trimmed}'. Use :help.");
                success = false;
                continue;
            }

            GameEventScriptInstance? instance = null;
            string? activePath = null;
            try
            {
                if (command == ":unload")
                {
                    session.Inventory.Unload(commandArgument);
                    if (!quiet) Console.Error.WriteLine($"Unloaded {commandArgument}.");
                    continue;
                }
                if (command is ":unloadAll" or ":reload")
                {
                    if (commandArgument.Length != 0) throw new ArgumentException($"{command} accepts no arguments.");
                    if (command == ":unloadAll")
                    {
                        var count = session.Inventory.UnloadAll();
                        if (!quiet) Console.Error.WriteLine($"Unloaded {count} programs. Native console handlers remain active.");
                    }
                    else
                    {
                        if (!session.Reload(ref activePath)) return false;
                        if (!quiet) Console.Error.WriteLine("Reloaded all active programs on a fresh host; initialization completed.");
                    }
                    continue;
                }
                if (command == ":dump")
                {
                    session.Inventory.WriteDump(commandArgument, color);
                    continue;
                }
                if (command == ":source")
                {
                    session.Inventory.WriteSource(commandArgument, color);
                    continue;
                }
                if (command == ":load")
                {
                    activePath = LoadPath(commandArgument);
                    session.Inventory.Load(RunProgramFiles.ReadOne(activePath), [activePath]);
                    if (!session.Pump()) return false;
                    if (!quiet) Console.Error.WriteLine($"Loaded {activePath}; initialization completed. Handlers remain active.");
                    continue;
                }
                var sourceName = $"<interactive:{lineNumber}>";
                var source = "on initialization {\n" + line + "\n}\n";
                var program = GameEventScriptBuilder.Create().AddScript(source, sourceName).Compile();
                var handlers = program.Bindings.Entries.Where(binding => binding.Kind is GameEventScriptBinaryBindKind.MessageHandler or GameEventScriptBinaryBindKind.MessageNameHandler).ToArray();
                if (handlers.Length != 1 || program.StringConstants.Resolve(handlers[0].Name) != "initialization")
                {
                    Console.Error.WriteLine("error cli.interactiveInput: Enter handler-body statements, not additional handler declarations.");
                    success = false;
                    continue;
                }
                instance = session.Host.Load(program);
                if (!session.Pump()) return false;
            }
            catch (GameEventScriptCompileException exception)
            {
                foreach (var diagnostic in exception.Diagnostics) RunObserver.WriteDiagnostic(diagnostic, "compile");
                success = false;
            }
            catch (GameEventScriptDynamicLinkException exception)
            {
                RunObserver.WriteDiagnostic(exception.Diagnostic, "link");
                success = false;
            }
            catch (GameEventScriptProgramFormatException exception)
            {
                RunObserver.WriteDiagnostic(exception.Diagnostic, activePath ?? "decode");
                if (exception.ByteOffset is { } offset) Console.Error.WriteLine($"  byteOffset={offset}");
                if (exception.SectionType is { } section) Console.Error.WriteLine($"  sectionType=0x{section:X4}");
                if (exception.EntryIndex is { } entry) Console.Error.WriteLine($"  entryIndex={entry}");
                success = false;
            }
            catch (DecoderFallbackException exception)
            {
                Console.Error.WriteLine($"{activePath}: error cli.invalidEncoding: {exception.Message}");
                success = false;
            }
            catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
            {
                Console.Error.WriteLine($"{activePath}: error cli.io: {exception.Message}");
                success = false;
            }
            catch (ArgumentException exception)
            {
                Console.Error.WriteLine($"error cli.consoleCommand: {exception.Message}");
                success = false;
            }
            finally
            {
                instance?.Detach();
            }
        }
    }

    private static string LoadPath(string text)
    {
        if (text.Length > 0 && text[0] is '\'' or '"')
        {
            var quote = text[0];
            if (text.Length < 2 || text[^1] != quote) throw new ArgumentException("The quoted :load path must end with a matching quote.");
            text = text[1..^1].Replace(new string(quote, 2), quote.ToString(), StringComparison.Ordinal);
        }
        if (string.IsNullOrWhiteSpace(text)) throw new ArgumentException("Specify one file after :load. Use :help load for examples.");
        return text;
    }
}
